lib/pbench/server/api/resources/generate_key.py

```diff
@@ -0,0 +1,147 @@
+from datetime import datetime, timedelta, timezone
+from http import HTTPStatus
+
+from flask import current_app, jsonify, make_response
+from flask.wrappers import Request, Response
+from flask_restful import abort
+import jwt
+from sqlalchemy.exc import IntegrityError, SQLAlchemyError
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.resources import (
+    APIAbort,
+    ApiAuthorizationType,
+    ApiBase,
+    ApiContext,
+    ApiMethod,
+    ApiParams,
+    ApiSchema,
+)
+import pbench.server.auth.auth as Auth
+from pbench.server.database.models.api_key import APIKey
+from pbench.server.database.models.audit import OperationCode
+
+
+def generate_key(time_delta: timedelta, user):
+    user_obj = user.get_json()
+    current_utc = datetime.now(timezone.utc)
+    expiration = current_utc + time_delta
+    payload = {
+        "iat": current_utc,
+        "exp": expiration,
+        "user_id": user_obj["id"],
+        "username": user_obj["username"],
+        "audience": Auth.oidc_client.client_id,
+    }
+    try:
+        generated_api_key = jwt.encode(
+            payload, current_app.secret_key, algorithm=Auth._TOKEN_ALG_INT
+        )
+    except (
+        jwt.InvalidIssuer,
+        jwt.InvalidIssuedAtError,
+        jwt.InvalidAlgorithmError,
+        jwt.PyJWTError,
+    ):
+        current_app.logger.exception(
+            "Could not encode the JWT api_key for user: {}", user
+        )
+        abort(
```

Comment:
Another thing, I am not sure if this `abort` will work because Flask catches exceptions raised by the application and handles them internally, instead of propagating them up the call stack. This means that if you call abort() inside an exception block, Flask may not see the exception raised by `abort` and may not be able to handle it properly (it would just return 500). In this case you are already rasing 500 so you don't need an abort call at all.

Reply:
@npalaska, your point is excellent (and it's part of the basis of our policy of not calling `abort()` from subroutines), but I think you have this backwards:

> Flask catches exceptions raised by the application and handles them internally, instead of propagating them up the call stack.

I believe that the up-stack code _is_ Flask, and the concern is that one of our `except Exception` blocks will catch the exception raised by `abort()` and "do the wrong thing" with it.  (This is part of the reason why `except Exception` blocks are not a good idea....)

> you are already raising 500 so you don't need an abort call at all.

I think this code is catching a JWT exception and trying to report the error.  It's true that we could let the exception propagate out and Flask would turn it into a `500` for us, but we really don't want that:  instead, we want to raise an `APIInternalError` instance, because that does useful stuff like creating an error ID, logging it, and reporting it out to the user.